sql/stats: TestCreateStatisticsCanBeCancelled failed

sql/stats.TestCreateStatisticsCanBeCancelled [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11380706?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11380706?buildTab=artifacts#/) on master @ [99a37c5b3a84aa25f6fb5232ddb5a42899325157](https://github.com/cockroachdb/cockroach/commits/99a37c5b3a84aa25f6fb5232ddb5a42899325157):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 63080 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* runtime/pprof.Do({0x5c4bed8?, 0x40001b2000?}, {{0x40001cb6e0?, 0x0?, 0x0?}}, 0x4006372f98)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x4006372f58?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:344 +0x6c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:336 +0xf0
* 
* goroutine 60530 [select, 2 minutes]:
* github.com/cockroachdb/cockroach/pkg/util/log.(*fileSink).gcDaemon(0x4009fce750, {0x5c4bea0, 0x4009bb4280})
* 	github.com/cockroachdb/cockroach/pkg/util/log/file_log_gc.go:25 +0x70
* created by github.com/cockroachdb/cockroach/pkg/util/log.ApplyConfig
* 	github.com/cockroachdb/cockroach/pkg/util/log/flags.go:319 +0xec8
* 
* goroutine 60615 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x5c4bf48, 0x40052704b0})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:346 +0x80
* runtime/pprof.Do({0x5c4bed8?, 0x40001b2000?}, {{0x40001cb6e0?, 0x5c4bf48?, 0x0?}}, 0x4005bfef98)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x0?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:344 +0x6c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:336 +0xf0
* 
* goroutine 60531 [select, 2 minutes]:
* github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).startTickerLocked.func1()
* 	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:575 +0xbc
* created by github.com/cockroachdb/pebble/vfs.(*diskHealthCheckingFS).startTickerLocked
* 	github.com/cockroachdb/pebble/vfs/external/com_github_cockroachdb_pebble/vfs/disk_health.go:564 +0x84
* 
* goroutine 60562 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x5c4bf48, 0x400cf528d0})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:346 +0x80
* runtime/pprof.Do({0x5c4bed8?, 0x40001b2000?}, {{0x40001cb6e0?, 0x5c4bf48?, 0x400ce8d750?}}, 0x40077ec798)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x40077ec758?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:344 +0x6c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:336 +0xf0
* 
* goroutine 60561 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x5c4bf48, 0x4004036480})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:346 +0x80
* runtime/pprof.Do({0x5c4bed8?, 0x40001b2000?}, {{0x40001cb6e0?, 0x400cad6f00?, 0x0?}}, 0x40037cbf98)
* 	GOROOT/src/runtime/pprof/runtime.go:40 +0x80
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0x400ab6eb40?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:344 +0x6c
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:336 +0xf0
* 
*
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCreateStatisticsCanBeCancelled.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30738

--- a/pkg/sql/stats/create_stats_job_test.go
+++ b/pkg/sql/stats/create_stats_job_test.go
@@ -30,7 +30,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/jobutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
@@ -128,7 +127,6 @@ func TestCreateStatsControlJob(t *testing.T) {
 func TestCreateStatisticsCanBeCancelled(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	skip.UnderStress(t, "test can be slow to quiesce because of filter")
 
 	var allowRequest chan struct{}
 
@@ -145,13 +143,13 @@ func TestCreateStatisticsCanBeCancelled(t *testing.T) {
 	sqlDB := sqlutils.MakeSQLRunner(conn)
 
 	sqlDB.Exec(t, `CREATE DATABASE d`)
-	sqlDB.Exec(t, `CREATE TABLE d.t (x INT PRIMARY KEY)`)
+	sqlDB.Exec(t, `CREATE TABLE d.t (x INT PRIMARY KEY)  WITH (sql_stats_automatic_collection_enabled = false)`)
 	sqlDB.Exec(t, `INSERT INTO d.t SELECT generate_series(1,1000)`)
 	var tID descpb.ID
 	sqlDB.QueryRow(t, `SELECT 'd.t'::regclass::int`).Scan(&tID)
 	setTableID(tID)
 
-	// Run CREATE STATISTICS and wait for to create the job.
+	// Run CREATE STATISTICS and wait for it to create the job.
 	allowRequest = make(chan struct{})
 	errCh := make(chan error)
 	go func() {
@@ -159,6 +157,7 @@ func TestCreateStatisticsCanBeCancelled(t *testing.T) {
 		errCh <- err
 	}()
 	allowRequest <- struct{}{}
+	setTableID(descpb.InvalidID)
 	testutils.SucceedsSoon(t, func() error {
 		row := conn.QueryRow("SELECT query_id FROM [SHOW CLUSTER STATEMENTS] WHERE query LIKE 'CREATE STATISTICS%';")
 		var queryID string
@@ -168,9 +167,22 @@ func TestCreateStatisticsCanBeCancelled(t *testing.T) {
 		_, err := conn.Exec("CANCEL QUERIES VALUES ((SELECT query_id FROM [SHOW CLUSTER STATEMENTS] WHERE query LIKE 'CREATE STATISTICS%'));")
 		return err
 	})
-	err := <-errCh
-	allowRequest <- struct{}{}
-
+	// Allow the filter to pass everything until an error is received.
+	var err error
+	testutils.SucceedsSoon(t, func() error {
+		// Assume something will fail.
+		err = errors.AssertionFailedf("failed for create stats to cancel")
+		for {
+			select {
+			case err = <-errCh:
+				return nil
+			case allowRequest <- struct{}{}:
+			default:
+				return err
+			}
+		}
+	})
+	close(allowRequest)
 	require.ErrorContains(t, err, "pq: query execution canceled")
 }
 
@@ -214,7 +226,6 @@ func TestAtMostOneRunningCreateStats(t *testing.T) {
 	case err := <-errCh:
 		t.Fatal(err)
 	}
-
 	autoStatsRunShouldFail := func() {
 		_, err := conn.Exec(`CREATE STATISTICS __auto__ FROM d.t`)
 		expected := "another CREATE STATISTICS job is already running"
@@ -534,7 +545,7 @@ func createStatsRequestFilter(
 		if req, ok := ba.GetArg(kvpb.Scan); ok {
 			_, tableID, _ := encoding.DecodeUvarintAscending(req.(*kvpb.ScanRequest).Key)
 			// Ensure that the tableID is what we expect it to be.
-			if tableID > 0 && descpb.ID(tableID) == tableToBlock.Load().(descpb.ID) {
+			if tableID > 0 && descpb.ID(tableID) == tableToBlock.Load() {
 				// Read from the channel twice to allow jobutils.RunJob to complete
 				// even though there is only one ScanRequest.
 				<-*allowProgressIota

